test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java

```diff
@@ -47,33 +43,26 @@ public TcpReadContext(TcpNioSocketChannel channel, TcpReadHandler handler, TcpFr
         this.handler = handler;
         this.channel = channel;
         this.frameDecoder = frameDecoder;
-        this.references.add(NetworkBytesReference.wrap(new BytesArray(new byte[DEFAULT_READ_LENGTH])));
     }
 
     @Override
     public int read() throws IOException {
-        NetworkBytesReference last = references.peekLast();
-        if (last == null || last.hasWriteRemaining() == false) {
-            this.references.add(NetworkBytesReference.wrap(new BytesArray(new byte[DEFAULT_READ_LENGTH])));
+        if (channelBuffer.getRemaining() == 0) {
+            channelBuffer.ensureCapacity(channelBuffer.getCapacity() + InboundChannelBuffer.PAGE_SIZE);
```

Comment:
I am not sure if `InboundChannelBuffer.PAGE_SIZE` should be externally visible. Usually the contract of `ensureCapacity` is to guarantee some kind of capacity. it's up the the impl to ensure that we overallocate with certain page sizes. I think this call should be like this `channelBuffer.ensureCapacity(channelBuffer.getCapacity() + 1)` and internally we can make sure we overallocate?

Reply:
then `InboundChannelBuffer.PAGE_SIZE` can be private